_docs/_latest/develop/dev/gcp-api.md

```diff
@@ -0,0 +1,213 @@
+---
+title: GCP API
+order: 103
+---
+
+# {{ page.title }}
+
+This page describes how to add a new GCP API client, which will allow
+Forseti to interact with GCP.
+
+---
+
+Before writing a new API client, it's important to note that GCP APIs have
+common methods:
+
+* list
+* get
+* create
+* update
+* delete
+
+Instead of writing these methods separately and repeatedly for all the
+different API clients, they can be written once as mixin classes, and then
+inherited by each API client. This makes it easy and quick to add new API
+clients, without having to rewrite these common methods for every API clients.
+
+You can [learn more about the mixin pattern here](https://www.ianlewis.org/en/mixins-and-python).
+
+You can see the currently available [mixins in Forseti here](https://github.com/GoogleCloudPlatform/forseti-security/blob/stable/google/cloud/forseti/common/gcp_api/repository_mixins.py).
+
+## Main Steps
+
+The main steps to add a new foo API client are:
+
+1. Define the API name & versions to be added
+1. Create a new foo API client file
+1. Create a FooClient class to provide the entry point to the GCP API method
+1. Create a FooRepositoryClient class, and define the property method for the resource that you want to interact with the GCP API
+1. Create a \_FooBarRepository class, to install the base GCP API functionalities (building requests, authentication, mixin classes)
+1. Use the new API client
+
+A good example to look at, that is small, complete, and self-contained would be
+[cloudsql.py](https://github.com/GoogleCloudPlatform/forseti-security/blob/stable/google/cloud/forseti/common/gcp_api/cloudsql.py).
+We will use it for the code-walk below.
+
+## Define the API name & versions to be added
+
+Add a new entry for the API and the versions that can be used in the
+`SUPPORTED_APIS` map.
+
+[google/cloud/forseti/common/gcp_api/_supported_apis.py](https://github.com/GoogleCloudPlatform/forseti-security/blob/stable/google/cloud/forseti/common/gcp_api/_supported_apis.py)
+
+```python
+SUPPORTED_APIS = {
+    'bigquery': {
+        'default_version': 'v2',
+        'supported_versions': ['v2']
+    },
+    'cloudresourcemanager': {
+        'default_version': 'v1',
+        'supported_versions': ['v1', 'v2']
+    },
+    'foo': {
+        'default_version': 'v1',
+        'supported_versions': ['v1', 'beta']
+    },
+    'sqladmin': {
+        'default_version': 'v1beta4',
+        'supported_versions': ['v1beta4']
+    },
+}
+```
+
+## Create a new foo API client file
+
+Create a new foo API client file called foo.py in
+[google/cloud/forseti/common/gcp_api](https://github.com/GoogleCloudPlatform/forseti-security/blob/stable/google/cloud/forseti/common/gcp_api)
+package.
+
+## Create a FooClient class to provide the entry point to the GCP API method
+
+1. Create FooClient class in foo.py.
+1. Initialize with the allowed quota and FooRepositoryClient.
+1. Create a get_foos(), which will invoke the actual GCP API
```

Comment:
Instead of get_foos, it should really be get_(repository_client_property) e.g. get_instances if the repos client has property instances. It can also be list if the repos has list mixin.